add batch docker dependency

cseed opened the pull request:
Shit, one more fix needed to get deploy back on track.  Missed in this change: https://github.com/hail-is/hail/pull/5644

cseed commented:
Hmm, there's another issue with our dependency management.  If batch changes and docker doesn't, docker won't get rebuilt locally and batch will fail.  I made docker touch a build stamp file (so we don't end up building docker multiple times) and make scorecard and batch build it before building themselves.

cseed commented on batch/hail-ci-build.sh:
```diff
@@ -1,9 +1,6 @@
 conda activate hail-batch
 
-flake8 batch
-pylint batch --rcfile batch/pylintrc --score=n
```
These seem to be duplicates because test-local-in-cluster depends on $(PY_CHECKERS)?

danking replied:
👍

cseed commented:
OK, I made a suite of additional changes:
 - create docker/requirements.txt,
 - batch doesn't use conda,
 - pr-builder (rebuilt) installs docker/requirements.txt (same requirements as base image),
 - put Spark in base image, removed spark-base image,
 - set ENV IN_HAIL_CI=1 in hail-ci scripts,
 - pull remote images and use --cache-from, allow push and deploy in CI, build locally only otherwise

cseed commented:
Rebuilding the pr-builder.  I had to switch from Debian to Ubuntu to pick up Python 3.6 by default.  Hopefully it will be done by morning.  Ha!

cseed commented:
OK, pr-builder image successfully rebuilt.  This should be ready for a look.

danking commented:
@cseed I added inter-project dependencies to `projects.yaml` does that address the docker-batch dependence?

danking commented on batch/Makefile:
```diff
@@ -1,30 +1,27 @@
-	docker build -t batch .
+	make -C ../docker build
+	-docker pull gcr.io/$(PROJECT)/batch
+	docker build -t batch --cache-from batch,base,ubuntu:18.04 .
```
We should also push the `latest` tag on every deploy.

In general, I prefer explicit tags in image names because it's non-obvious that `docker pull gcr.io/$(PROJECT)/batch` pulls the `latest` tag. Even with that knowledge, `latest` is a misleadingly named tag. It really ought to be called `default`, it has no relation to the most recently pushed or most recently built image.

danking replied:
And to reiterate: in current docker, this means if you change layer `k`, successfully build layers `k` to `k'`, and fail on layer `k'+1`, you will have to rebuild layers `k` to `k'` because your `--cache-from` does not include the anonymous, temporary images for `k` through `k'`.

cseed replied:
Agreed, explicit is better, I'll put latest everywhere.

Yes.  As far as I can tell, there's no way to cache from partial failed local builds and a pre-existing remote build.  Which sucks balls when you're trying to iterate on a big Dockerfile like pr-builder.

danking commented on batch/Makefile:
```diff
@@ -44,27 +41,27 @@ push-test: build-test
+test-local-in-cluster: check
+	POD_NAMESPACE='test' ./test-locally.sh
 
 deploy: push
```
concretely, I'm suggesting that either `deploy` does a `docker push batch gcr.io/$(PROJECT)/batch:latest`

jigold commented on Dockerfile.pr-builder:
```diff
@@ -1,23 +1,23 @@
     cmake \
-    docker-ce \
+    curl \
+    docker.io \
```
Might be worth adding `jq` while you're at it.

jigold replied:
Actually, this is the wrong docker file.

cseed replied:
Oh, god, please don't ask me to change the pr-builder image again!  But yes, let's save it for the next update.

danking commented on batch/batch/server/server.py:
```diff
@@ -301,7 +301,8 @@ def __init__(self, pod_spec, batch_id, attributes, callback, parent_ids,
 
-    def refresh_parents_and_maybe_create(self):
+    # pylint incorrect error: https://github.com/PyCQA/pylint/issues/2047
+    def refresh_parents_and_maybe_create(self):  # pylint: disable=invalid-name
```
you could also redefine the function name regex in pylintrc. I somewhat prefer that. It's called `function-rgx`.

danking commented on docker/Makefile:
```diff
@@ -1,21 +1,32 @@
-SPARK_BASE_IMAGE = gcr.io/$(PROJECT)/spark-base:$(shell docker images -q --no-trunc spark-base:latest | sed -e 's,[^:]*:,,')
 
-build:
+build-stmp: Dockerfile.base
```
It also depends on every file in the context. `$(shell find . -type f)` should do it.

danking commented on docker/Makefile:
```diff
@@ -1,21 +1,32 @@
+.PHONY: push deploy
 
+PROJECT = $(shell gcloud config get-value project)
+BASE_LATEST = gcr.io/$(PROJECT)/base
```
I prefer `:latest`

danking commented on scorecard/Makefile:
```diff
@@ -5,8 +5,9 @@ PROJECT = $(shell gcloud config get-value project)
+	make -C ../docker build
 	-docker pull gcr.io/$(PROJECT)/scorecard
-	docker build . -t scorecard --cache-from scorecard,base,ubuntu:18.04
+	docker build -t scorecard --cache-from scorecard,base,ubuntu:18.04 .
```
This should probably include `gcr.io/$(PROJECT)/scorecard:latest` in preference to `scorecard` (which almost certainly is unavailable on the build machine). `deploy` should push latest.

danking commented:
@cseed I'm also OK approving and addressing these issues in another PR if that unblocks our deploy.

cseed commented:
OK, I think I addressed the comments:
 - I'm going to leave the pylint fix for another PR,
 - use latest explicitly for remote repositories,
 - explicit dependencies on build-stmp instead of find

cseed commented:
Whoa, it worked.

I included one change that might have warranted re-review.  I was getting errors becomes some Jobs, on which delete had been called, were still being used.  I tracked it down to a recent cancel => delete change in `PR.update_from_completed_batch_job`.  If look at that function, it is clear delete is not OK because in several cases the build object keep a handle to the job.  I reverted it, and now clear all the fields of Job when it is deleted.

https://github.com/hail-is/hail/pull/5655/files#diff-433f83d97fa8a526a3f8cff52590e422R479
https://github.com/hail-is/hail/pull/5655/files#diff-0c1f876ad25335b076837f768f727566R59

danking commented:
What were the errors? It should be OK to hold the job object around and e.g. use it to ask for logs even if the job is deleted.

danking commented:
I'm worried that we're back into a bad state where we'll keep seeing the same job over and over again and not realize we've already learned from it.

danking commented:
This broke CI, it can't update from batch

cseed commented:
> What were the errors? It should be OK to hold the job object around and e.g. use it to ask for logs even if the job is deleted.

We were setting the Job id attribute to None on deletion.  ci then queried various properties of the job, passing None as the job id in the URL.  That caused batch to 500 converting the id to an integer (this is itself a bug).  Leaving the id in the Job might also fix it, but it seems wrong in the REST setting to delete an object, but still be able to query it.
